File: 0022-generate-parentheses/solution.cpp
class Solution {
    void solve(int open, int close, string temp, vector<string> &ans){
        //base case 
        if(open == 0 && close == 0){
            ans.push_back(temp);
            return;
        }

        if(open){
            temp.push_back('(');
            solve(open-1, close, temp, ans);
            temp.pop_back();
        }
        if(close > open){
            temp.push_back(')');
            solve(open, close-1, temp, ans);
            temp.pop_back();
        }
    }
public:
    vector<string> generateParenthesis(int n) {
        vector<string> ans;
        string temp= "";

        solve(n, n, temp, ans);
        return ans;
    }
};

File: 0022-generate-parentheses/solution.java
class Solution {
    public void helper(int open, int close, String s, List<String> ans){
        if(open == 0 && close == 0){
            ans.add(s);
            return ;
        } 
        

        if(open != 0){
            s+="(";
            helper(open - 1, close, s, ans);
            s =  s.substring(0, s.length()-1);
        }

        if(open < close){
            s += ")";
            helper(open, close-1, s, ans);
            s = s.substring(0, s.length()-1);
        }
    }
    public List<String> generateParenthesis(int n) {
       List<String> ans = new ArrayList<String>(); // ✅
        String s = "";
        helper(n, n, s, ans);
        return ans;
    }
}
